### PR TITLE
[C++] Fix bugs in UnbufferedCharStream

### PR DESCRIPTION
**Pull request by jcking**
As the name says, fix bugs in UnbufferedCharStream pointed out by #3412. I also cleaned up the class while I was there.

### CODE REVIEW

**Comment by jcking** [comment]
@mike-lischke

**Comment by jcking** [comment]
I also removed `virtual` prefixes where it is overriding, because it is unnecessary as override implies the method was virtual and will fail it it was not.

**Comment by jcking** [comment]
I also chose to make `toString` throw, because it is not implemented in the Java version and would just return the Java representation which is usually the class name and an address/hash.

File: runtime/Cpp/runtime/src/UnbufferedCharStream.cpp [hunk 1]
```diff
@@ -101,7 +98,7 @@ size_t UnbufferedCharStream::LA(ssize_t i) {
     return EOF;
   }
 
-  if (_data[static_cast<size_t>(index)] == 0xFFFF) {
+  if (_data[static_cast<size_t>(index)] == WEOF) {
```
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/UnbufferedCharStream.cpp:101]: What is WEOF?

Comment (jcking) [reply runtime/Cpp/runtime/src/UnbufferedCharStream.cpp]: WEOF is the wide character version of EOF, defined in wchar.h. I switched it to `std::char_traits<wchar_t>::eof()` which is the preferred C++ equivalent.

**Review by mike-lischke — APPROVED** [review]
@parrt This C++ patch can be merged.

**Pull request merged**